app/Concerns/HasRatings.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+
+namespace App\Concerns;
+
+use App\Events\User\Updated;
+use App\Libraries\CERT\VATSIMUserDetails;
+use App\Models\Rating;
+use App\Models\RatingPivot;
+use Carbon\Carbon;
+use Illuminate\Database\Eloquent\Relations\BelongsToMany;
+
+trait HasRatings
+{
+    public function ratings(): BelongsToMany
+    {
+        return $this->belongsToMany(
+            Rating::class,
+            "user_ratings",
+            'user_id',
+            'rating_id'
+        )->using(RatingPivot::class)
+            ->wherePivot('deleted_at', '=', null)
+            ->withPivot('deleted_at')
+            ->withTimestamps();
+    }
+
+    public function syncRatings(int $atcRatingCode, int $pilotRatingCode): void
+    {
+        $ratings = collect();
+
+        // Handle ATC rating
+        if ($atcRatingCode === 0) {
+            $this->banNetwork('Network ban discovered via Cert login.');
+            return;
+        }
+
+        // Check if has non-ATC primary rating
+        if ($atcRatingCode >= 8) {
+            $previousRating = VATSIMUserDetails::getPreviousRatingsInfo($this->id);
+            if ($previousRating && isset($previousRating->PreviousRatingInt) && $previousRating->PreviousRatingInt > 0) {
+                // Push pre-special rating
+                $ratings->push(Rating::atcRatingFromID($previousRating->PreviousRatingInt));
+            }
+            // Check if we are going to be adding a different "special" rating than current
+            if($this->specialRating->id != $atcRatingCode){
+                $pivot = $this->specialRating->pivot;
+                $pivot->deleted_at = Carbon::now();
+                $pivot->save();
+            }
+        } else {
+            // Remove any "special" ATC ratings
+            if($this->specialRating){
+                $pivot = $this->specialRating->pivot;
+                $pivot->deleted_at = Carbon::now();
+                $pivot->save();
+            }
+        }
+
+        $this->endNetworkBanIfHas();
+        // Push current rating
+        $ratings->push(Rating::atcRatingFromID($atcRatingCode));
+
+        // Handle pilot ratings
+        for ($i = 1; $i <= 256; $i *= 2) {
+            if ($i & $pilotRatingCode) {
+                $ratings->push(Rating::typePilot()->networkValue($i)->first());
+            }
+        }
```

Comment:
magic numbers make this really hard to understand what's going on - maybe a candidate for a private function

Reply:
tbh I don't really know what happens here either - also stolen from Core code...

Reply:
someone likes living dangerously

Reply:
I think this is partially an issue due to how the `vatsim_id` of the pilot ratings is calculated...

I don't think abstracting to a private function would really help. Any ideas or happy to leave be till CERT v2 :O